File: clients/cpp_clang/model/Quest.cpp
#include "Quest.hpp"

namespace model {

Quest::Quest(model::Vec2Int pickupCell, model::Vec2Int dropCell, long long score) : pickupCell(pickupCell), dropCell(dropCell), score(score) { }

// Read Quest from input stream
Quest Quest::readFrom(InputStream& stream) {
    model::Vec2Int pickupCell = model::Vec2Int::readFrom(stream);
    model::Vec2Int dropCell = model::Vec2Int::readFrom(stream);
    long long score = stream.readLongLong();
    return Quest(pickupCell, dropCell, score);
}

// Write Quest to output stream
void Quest::writeTo(OutputStream& stream) const {
    pickupCell.writeTo(stream);
    dropCell.writeTo(stream);
    stream.write(score);
}

// Get string representation of Quest
std::string Quest::toString() const {
    std::stringstream ss;
    ss << "Quest { ";
    ss << "pickupCell: ";
    ss << pickupCell.toString();
    ss << ", ";
    ss << "dropCell: ";
    ss << dropCell.toString();
    ss << ", ";
    ss << "score: ";
    ss << score;
    ss << " }";
    return ss.str();
}

bool Quest::operator ==(const Quest& other) const {
    return pickupCell == other.pickupCell && dropCell == other.dropCell && score == other.score;
}

}

size_t std::hash<model::Quest>::operator ()(const model::Quest& value) const {
    size_t result = 0;
    result ^= std::hash<model::Vec2Int>{}(value.pickupCell) + 0x9e3779b9 + (result << 6) + (result >> 2);
    result ^= std::hash<model::Vec2Int>{}(value.dropCell) + 0x9e3779b9 + (result << 6) + (result >> 2);
    result ^= std::hash<long long>{}(value.score) + 0x9e3779b9 + (result << 6) + (result >> 2);
    return result;
}

File: clients/python_pypy/model/quest.py
from model.vec2_int import Vec2Int
from stream_wrapper import StreamWrapper

class Quest:
    """TODO - Document"""

    __slots__ = ("pickup_cell","drop_cell","score",)

    pickup_cell: Vec2Int
    drop_cell: Vec2Int
    score: int

    def __init__(self, pickup_cell: Vec2Int, drop_cell: Vec2Int, score: int):
        self.pickup_cell = pickup_cell
        """TODO - Document"""
        self.drop_cell = drop_cell
        """TODO - Document"""
        self.score = score
        """TODO - Document"""

    @staticmethod
    def read_from(stream: StreamWrapper) -> "Quest":
        """Read Quest from input stream
        """
        pickup_cell = Vec2Int.read_from(stream)
        drop_cell = Vec2Int.read_from(stream)
        score = stream.read_long()
        return Quest(pickup_cell, drop_cell, score)
    
    def write_to(self, stream: StreamWrapper):
        """Write Quest to output stream
        """
        self.pickup_cell.write_to(stream)
        self.drop_cell.write_to(stream)
        stream.write_long(self.score)
    
    def __repr__(self):
        return "Quest(" + \
            repr(self.pickup_cell) + \
            ", " + \
            repr(self.drop_cell) + \
            ", " + \
            repr(self.score) + \
            ")"